h2o-py/h2o/display.py

```diff
@@ -6,126 +6,662 @@
+        """
+        with repr_context('pretty'):
+            return str(self)
+
+    def _str_(self):
+        """Override this method to return the informal string representation."""
+        return repr(self)
+
+    def _str_html_(self):
+        """Override this method to return a string description in html format."""
+        return self._str_()
+
+    def _str_pretty_(self):
+        """Override this method to return a pretty string description used by default as repl output."""
+        return self._str_()
+
+
+class H2ODisplay(DisplayMixin):
+    """
+    A convenient mixin for H2O classes, providing standard methods for formatting and rendering.
+    """
+
+    def show(self, fmt=None):
+        """
+        Describe and renders the current object in the given format,
+        by default guessing the best format for the current environment.
+        
+        :param fmt: one of (None, 'auto', 'plain', 'pretty', 'html').
+                    Defaults to None/'auto'.
+        """
+        display(self, fmt=fmt)
+        
+    def to_html(self):
+        """
+        :return: a html representation of the current object.
+        """
+        return to_html(self)
+    
+    def to_pretty_str(self):
+        """
+        :return: a pretty string representation of the current object.
+        """
+        return to_pretty_str(self)
+    
+    def to_str(self, fmt=None):
+        """
+        :return: a string representation of the current object.
+        :param fmt: one of (None, 'plain', 'pretty', 'html').
+                    Defaults to None/'plain'.
+        """
+        return to_str(self, fmt=fmt)
+
+
+class H2OStringDisplay(H2ODisplay):
+    """
+    Wrapper ensuring that the given string is rendered consistently in unique format for all environments.
+    
+    """
+    
+    def __init__(self, s):
+        self._s = s
+        
+    def _repr_(self):
+        return self._s
+    
+
+class H2ODisplayWrapper(H2ODisplay):
+    """
+    Wraps a function returning a string into a displayable object 
+    that will call the function with the requested format, depending on the environment.
+    """
+
+    def __init__(self, repr_fn):
+        """
+        :param repr_fn: the wrapped representation function with signature: (fmt: Optional[str]) -> str
+        See ``display`` function for the list supported formats. 
+        """
+        self._repr_fn = repr_fn
+
+    def _str_(self):
+        return self._repr_fn()
+
+    def _str_pretty_(self):
+        return self._repr_fn(fmt='pretty')
+
+    def _str_html_(self):
+        return self._repr_fn(fmt='html')
+
+
+class H2OItemsDisplay(H2ODisplay):
+
+    def __init__(self, items):
+        """
+        :param items: a list of items to be rendered 
+        """
+        self._items = items
+
+    def _str_(self):
+        return format_to_multiline(self._items)
+
+    def _str_html_(self):
+        return format_to_html(self._items)
+
+
+class H2OTableDisplay(H2ODisplay):
```

Comment:
I need to check what engine DataSpell is using: even in basic python REPL, the basic rendering is better than what you're showing on top… so I don't know what's happening.

Reply:
didn't change the float representation from what it used to be, will see if I can improve that easily

Reply:
also with the version you were testing, it should have tried to use pandas by default, which was a change from legacy behaviour. As it broke several tests, I'm now considering using our table representation by default.

Reply:
Thanks. That appears to fix that issue.
<img width="1069" alt="Screen Shot 2022-06-30 at 12 06 12" src="https://user-images.githubusercontent.com/61695433/176652738-ea396f16-fdac-4fd1-bf2d-e03a6301823e.png">

<img width="786" alt="Screen Shot 2022-06-30 at 12 13 03" src="https://user-images.githubusercontent.com/61695433/176652883-230a2d8c-af6c-4e9d-a935-dbba2974278d.png">

Reply:
I tested PyCharm now and there is an issue with H2OFrame (leaderboard) - especially the last two columns.

<img width="1456" alt="Screen Shot 2022-06-30 at 17 19 17" src="https://user-images.githubusercontent.com/61695433/176715905-4177a4bb-24fc-4654-b319-39f58102f3e3.png">

Jupyter looks still good:

<img width="1296" alt="Screen Shot 2022-06-30 at 17 23 39" src="https://user-images.githubusercontent.com/61695433/176716194-70a9ec4b-0a16-420c-b2d1-a9a21843a2ff.png">

The previous example looks the same way as in DataSpell so I believe PyCharm and DataSpell use the same thing to render jupyter.
<img width="1182" alt="Screen Shot 2022-06-30 at 17 19 25" src="https://user-images.githubusercontent.com/61695433/176715778-77802271-f6cc-4379-be76-d18a2056ae6b.png">

Otherwise, I tested also spyder and that seemed to work ok (pretty printing to the console).

Reply:
for JetBrains tools, I fixed the rendering of H2OFrame, but I won't do the same for those "smaller" TwoDimTables as our `show` functions sometimes render several of them and if I fix them to be detected by DataSpell as `dataframes` (using the `class="dataframe" fix-up) then the renderer only renders those and skips the rest, which is even worse.
Ideally, if I could detect those JB tools, I would just render the smaller tables using `pretty` format instead of HTML as it looks better there.

Also, note that I removed the gray background: there's now a border below the header row that is just enough.
I think I'm done with the cosmetics for this PR: I'm already glad that a few things look already better, there's plenty of time to improve formatting later if needed.

Thanks for your extensive testing on this.